### ISSUE TITLE
createEntity could return a stronger type

### ISSUE THREAD

**Issue by mikecann** [hmans/miniplex#156]
Because after you `createEntity` you guarantee that certain components are going to be there you could return the type from `createEntity` with "required" components.

For now I am hacking this with an "as":

```ts
export const createCell = ({
  resources,
  world,
  overrides,
}: {
  resources: ResourceCache;
  world: GameWorld;
  overrides?: ComponentOverrides;
}) => {
  return world.createEntity({
    position: { x: 0, y: 0, z: 0 },
    mesh: { mesh: resources.cellMesh.clone() },
    ...overrides,
  }) as EntityWith<"position" | "mesh">;
};
```

where:

```ts
import { World } from "miniplex";
import { Camera, FreeCamera, Mesh } from "@babylonjs/core";

export interface Entity {
  position: { x: number; y: number; z: number };
  velocity?: { x: number; y: number; z: number };
  mesh?: { mesh: Mesh };
  camera?: { camera: FreeCamera };
}

export type ComponentOverrides = Partial<Entity>;

export type ComponentNames = keyof Entity;

export type EntityWith<TComponentNames extends ComponentNames> = Required<
  Pick<Entity, TComponentNames>
>;
```

Thoughts?

**Comment by hmans** [comment]
Thanks for the issue! I assume we're talking about the current state of the `main` branch, which is preparing for a 1.0 release.

I was inferring the type from the entity passed into `createEntity` until recently, but this had the problem of bypassing the intended type _checking_ (ie. it would allow props to be passed in that were not valid according to the world's entity type.) The upcoming `satisfies` feature in TypeScript 4.9 will help here, and I'm tempted to migrate to TS 4.9 and use that before pushing the final 1.0 version.

(If you have other ideas, please let me know!)

**Comment by mikecann** [comment]
Yes I was referring to whatever the latest version is on the main branch. 

>  but this had the problem of bypassing the intended type checking (ie. it would allow props to be passed in that were not valid according to the world's entity type.)

Yes I had a brief go at trying to do a PR for this by taking in a definable type into `createEntity` but couldnt get it to work:

```ts
/**
 * A RegisteredEntityWith is a RegisteredEntity (see above) but is also guaranteed to have certain components
 */
export type RegisteredEntityWith<
  TEntity extends IEntity,
  TComponentNames extends keyof TEntity
> = Required<Pick<TEntity, TComponentNames>> & MiniplexComponent<TEntity>

public createEntity<TInput extends T>(entity: TInput): RegisteredEntityWith<T, keyof TInput> { 
...
```

But I was getting this error:

![image](https://user-images.githubusercontent.com/215033/193511592-dd4152f3-6b22-4b50-86b0-1c20aecbc352.png)

As you say, maybe satisfies in 4.9 will fix it.

**Comment by hmans** [comment]
I've done some experimenting. Not sure if `satisfies` will help fix it, but I will keep an eye on it (and also for other potential solutions.) I do want to solve this post 1.0, though. We'll figure it out :)

**Comment by hmans** [comment]
A quick update, this is coming with the upcoming Miniplex 2.0 release (beta soon.)

It is still unclear if TS will still throw a warning if extra properties are added when the type _extends_ the world entity type; [some recent experiments with this have shown very weird results](https://twitter.com/hmans/status/1582279679766982657) that don't make it obvious what the _intended_ behavior of TS is.

Having said that, even if TS allows extra props to live on a type that extends another type, losing his kind of checking against extra props is an acceptable compromise if it means `createEntity`'s return value will be more accurate. It is also in line with Miniplex 2.0's overall theme of being a lot more relaxed in a bunch of places, and easier to use.

So, I'm adding this issue to the "2.0" milestone and closing it. Thanks!

**Comment by hmans** [comment]
A quick update: it's been confirmed that TS should _not_ display a warning for extra props, and that the intermittent instances of it still doing so [is a bug](https://github.com/microsoft/TypeScript/issues/50818).

Still, I'm going to go ahead and ship this change with 2.0. It will mean that the objects returned by `add` (the new API for `createEntity`) has the correct type, but also that `add` will not complain about extra entities on the object. Which is simultaneously _good_ because it matches how TS typically thinks about types, and it shouldn't be _wrong_ to make an object an entity even if it has additional props not known to the World's entity type; but it's also _bad_ because now you can have a typo in a component name and not get a warning about it. But I think it's an acceptable compromise to make.
